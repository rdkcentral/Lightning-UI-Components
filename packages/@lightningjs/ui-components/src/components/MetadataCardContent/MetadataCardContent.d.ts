/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import lng from '@lightningjs/core';
import MetadataBase, { MetadataBaseStyle } from '../MetadataBase';
import Provider, { ProviderStyle } from '../Provider';
import TextBox from '../TextBox';
import type TextBoxStyle from '../TextBox/TextBox.style';
import { StylePartial } from '../../types/lui';

type MetadataCardContentStyle = MetadataBaseStyle & {
  detailsTextStyle: TextBoxStyle;
  descriptionTextStyle: TextBoxStyle;
  fadeWidth: number;
  provider: ProviderStyle;
};

declare namespace MetadataCardContent {
  export interface TemplateSpec extends MetadataBase.TemplateSpec {
    /**
     * text of title content
     */
    title?: string;
    /**
     * description text below title
     */
    description?: string;
    /**
     * details text directly below description
     */
    descriptionDetails?: string;
    /**
     * details text at bottom left of component
     */
    details?: string;
    /**
     * Object containing all properties supported in the Provider component
     */
    Provider?: lng.Element.PatchTemplate<Provider.TemplateSpec>;
  }
}

declare class MetadataCardContent<
  TemplateSpec extends MetadataBase.TemplateSpec = MetadataBase.TemplateSpec,
  TypeConfig extends MetadataBase.TypeConfig = MetadataBase.TypeConfig
> extends MetadataBase<TemplateSpec, TypeConfig> {
  /**
   * text of title content
   */
  title?: string;

  /**
   * description text below title
   */
  description?: string;
  /**
   * details text directly below description
   */
  descriptionDetails?: string;
  /**
   * details text at bottom left of component
   */
  details?: string;
  /**
   * Object containing all properties supported in the Provider component
   */
  Provider?: lng.Element.PatchTemplate<Provider.TemplateSpec>;

  get style(): MetadataCardContentStyle;
  set style(v: StylePartial<MetadataCardContentStyle>);

  // Tags

  get _Text(): TextBox;
  get _Title(): TextBox;
  get _Description(): TextBox;
  get _DescriptionDetails(): TextBox;
  get _DetailsWrapper(): TextBox;
  get _DetailsFader(): TextBox;
  get _Details(): TextBox;
  get _Provider(): lng.Component;
}

export { MetadataCardContent as default, MetadataCardContentStyle };
