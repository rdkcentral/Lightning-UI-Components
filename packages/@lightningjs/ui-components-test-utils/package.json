{
  "name": "@lightningjs/ui-components-test-utils",
  "version": "1.3.0",
  "description": "Helpful Jest unit test utilities for LightningJS applications utilizing the Lightning-UI-Components package.",
  "repository": {
    "url": "https://github.com/rdkcentral/Lightning-UI-Components",
    "directory": "packages/@lightningjs/ui-components-test-utils"
  },
  "homepage": "https://rdkcentral.github.io/Lightning-UI-Components/",
  "type": "module",
  "exports": "./index.js",
  "files": [
    "index.js",
    "src/**"
  ],
  "scripts": {
    "build": "rm -rf ./dist && rollup --config ../../../rollup/rollup.config.js --name lightning-ui-components"
  },
  "peerDependencies": {
    "@lightningjs/core": "^2.9.0"
  },
  "devDependencies": {
    "@lightningjs/ui-components-theme-base": "workspace:^",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "jest-sinon": "^1.1.0",
    "jest-webgl-canvas-mock": "^0.2.3",
    "semantic-release": "^19.0.5",
    "semantic-release-monorepo": "^7.0.5",
    "sinon": "^15.2.0"
  },
  "dependencies": {
    "@lightningjs/ui-components": "workspace:^"
  },
  "publishConfig": {
    "access": "public"
  }
}
