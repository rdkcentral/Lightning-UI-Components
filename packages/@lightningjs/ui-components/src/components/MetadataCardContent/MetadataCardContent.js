/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import MetadataBase from '../MetadataBase';
import TextBox from '../TextBox';
import Provider from '../Provider';
import FadeShader from '../../shaders/FadeShader';
import * as styles from './MetadataCardContent.styles';

export default class MetadataCardContent extends MetadataBase {
  static get __componentName() {
    return 'MetadataCardContent';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    return {
      Text: {
        flex: { direction: 'column', justifyContent: 'flex-start' },
        Title: {
          type: TextBox
        },
        Description: {
          type: TextBox
        },
        DescriptionDetails: {
          type: TextBox
        }
      },
      DetailsWrapper: {
        DetailsFader: {
          Details: {
            mountY: 0.5,
            type: TextBox,
            signals: {
              textBoxChanged: '_resolveDetails'
            }
          }
        },
        Provider: {
          mount: 1,
          type: Provider,
          signals: {
            providerChanged: '_resolveProvider'
          }
        }
      }
    };
  }

  static get properties() {
    return [
      'description',
      'descriptionDetails',
      'details',
      'provider',
      'title'
    ];
  }

  static get tags() {
    return [
      'Text',
      {
        name: 'Title',
        path: 'Text.Title'
      },
      {
        name: 'Description',
        path: 'Text.Description'
      },
      {
        name: 'DescriptionDetails',
        path: 'Text.DescriptionDetails'
      },
      'DetailsWrapper',
      {
        name: 'DetailsFader',
        path: 'DetailsWrapper.DetailsFader'
      },
      {
        name: 'Details',
        path: 'DetailsWrapper.DetailsFader.Details'
      },
      {
        name: 'Provider',
        path: 'DetailsWrapper.Provider'
      }
    ];
  }

  _setDetails(details) {
    if (details) {
      this._detailsPromise = new Promise(resolve => {
        this._detailsPromiseResolver = resolve;
      });
    } else {
      this._detailsPromise = undefined;
    }
    return details;
  }

  _setProvider(provider) {
    if (provider) {
      this._providerPromise = new Promise(resolve => {
        this._providerPromiseResolver = resolve;
      });
    } else {
      this._providerPromise = undefined;
    }
    return provider;
  }

  async _update() {
    this._updateLines();
    this._updateProvider();
    await Promise.all(
      [this._detailsPromise, this._providerPromise].filter(Boolean)
    );
    this._updatePositions();
  }

  _updateLines() {
    this._Text.w = this.w;
    this._updateTitle();
    this._updateDescription();
    this._updateDescriptionDetails();
    this._updateDetails();
  }

  _updateDescription() {
    this._Description.patch({
      content: this.description,
      style: {
        textStyle: {
          ...this.style.descriptionTextStyle,
          wordWrap: true,
          wordWrapWidth: this._Text.w
        }
      }
    });
  }

  _updateDescriptionDetails() {
    this._DescriptionDetails.patch({
      content: this.descriptionDetails,
      style: {
        textStyle: {
          ...this.style.descriptionDetailsTextStyle,
          maxLines: 1,
          wordWrap: true,
          wordWrapWidth: this._Text.w
        }
      },
      y: this.style.descriptionDetailsStyle.paddingY
    });
  }

  _updateDetails() {
    const maxWidth = this._detailsMaxW;
    this._Details.patch({
      content: this.details,
      style: {
        textStyle: {
          ...this.style.detailsTextStyle,
          wordWrap: true,
          maxLines: 1, // sets default for details
          wordWrapWidth: maxWidth + this.style.fadeWidth / 2
        }
      }
    });
    if (this._Details.finalW > maxWidth) {
      this._DetailsFader.patch({
        w: maxWidth + this.style.fadeWidth / 2,
        shader: {
          type: FadeShader,
          positionLeft: 0,
          positionRight: this.style.fadeWidth
        }
      });
    } else {
      this._DetailsFader.shader = undefined;
    }
  }

  _resolveDetails() {
    this._detailsPromiseResolver && this._detailsPromiseResolver();
    this._updatePositions();
  }

  _updateProvider() {
    if (this.provider) {
      this._Provider.patch({
        mode: this.mode,
        ...this.provider,
        style: this.style.provider
      });
    }
  }

  _resolveProvider() {
    this._providerPromiseResolver && this._providerPromiseResolver();
    this._updatePositions();
  }

  _updatePositions() {
    this._Text.h = this._textH;
    this._Text.w = this._textW;

    this._DetailsWrapper.w = this._textW;
    this._DetailsWrapper.h = Math.max(this._providerH, this._Details.h);
    this._DetailsWrapper.y = this.h - this._DetailsWrapper.h;

    this._Details.y = this._DetailsWrapper.h / 2;

    this._Provider.x = this._DetailsWrapper.w - this._providerW;
    this._Provider.y = this._DetailsWrapper.h - this._providerH;
  }

  get _textW() {
    return this.w;
  }

  get _textH() {
    return this.h - this._providerH;
  }

  get _providerW() {
    return this._Provider.w;
  }

  get _providerH() {
    return this._Provider._Row._Items.h;
  }

  get _detailsMaxW() {
    return this.w - this._providerW - this.style.fadeWidth / 2;
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return (
      this._announce || [
        this._Title && this._Title.announce,
        this._Description && this._Description.announce,
        this._DescriptionDetails && this._DescriptionDetails.announce,
        this._Details && this._Details.announce,
        this._Provider && this._Provider.announce
      ]
    );
  }
}
