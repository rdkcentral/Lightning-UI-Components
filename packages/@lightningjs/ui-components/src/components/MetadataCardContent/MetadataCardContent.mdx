import { Canvas, Story } from '@storybook/blocks';
import MetadataCardContent from '.';
import Provider from '../Provider';
import xfinityProviderLogo from '../../assets/images/Xfinity-Provider-Logo-Square.png';

# MetadataCardContent

A component to format metadata on CardContent. Contains a [Provider](?path=/docs/components-provider--basic) to format [Icon(s)](?path=/docs/components-icon--png).

## Source

https://github.com/rdkcentral/Lightning-UI-Components/tree/develop/packages/%40lightningjs/ui-components/src/components/MetadataCardContent

## Usage

```js
import { MetadataCardContent } from '@lightningjs/ui-components';

class Basic extends lng.Component {
  static _template() {
    return {
      MetadataCardContent: {
        type: MetadataCardContent,
        w: 600,
        h: 250,
        title: 'Title',
        description: 'Description',
        details: 'Details',
        provider: {
          visibleCount: args.visibleCount,
          providers: Array(10).fill(xfinityProviderLogoSquare)
        }
      }
    };
  }
}
```

<Canvas>
  <Story id="components-metadatacardcontent--metadata-card-content" />
</Canvas>

## API

### Properties

| name        | type                                                  | required | default   | description                                                                               |
| ----------- | ----------------------------------------------------- | -------- | --------- | ----------------------------------------------------------------------------------------- |
| title       | string                                                | false    | undefined | title text                                                                                |
| description | string                                                | false    | undefined | description text directly below title                                                     |
| details     | string                                                | false    | undefined | details text at bottom left of component                                                  |
| provider    | [Provider](?path=/docs/components-provider--provider) | false    | undefined | an object of [Provider](?path=/docs/components-provider--provider) properties to patch in |

### Style Properties

| name             | type             | description                                                                              |
| ---------------- | ---------------- | ---------------------------------------------------------------------------------------- |
| detailsTextStyle | string \| object | text style for details                                                                   |
| providerStyle    | object           | style properties to pass along to the [Provider](?path=/docs/components-provider--basic) |

### Methods
