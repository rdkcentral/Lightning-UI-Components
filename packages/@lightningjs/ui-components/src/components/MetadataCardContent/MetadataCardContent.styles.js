/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export const base = theme => ({
  detailsTextStyle: theme.typography.body3,
  descriptionTextStyle: { maxLines: 3 },
  descriptionDetailsStyle: { paddingY: 5 },
  fadeWidth: theme.spacer.md * theme.spacer.md,
  provider: {
    itemSize: theme.spacer.xxxl + theme.spacer.md
  }
});

export const tone = theme => ({
  neutral: {
    detailsTextStyle: { textColor: theme.color.textNeutral },
    mode: {
      disabled: {
        detailsTextStyle: { textColor: theme.color.textNeutralDisabled }
      }
    }
  },
  inverse: {
    detailsTextStyle: { textColor: theme.color.textInverse },
    mode: {
      disabled: {
        detailsTextStyle: { textColor: theme.color.textNeutralDisabled }
      }
    }
  },
  brand: {
    detailsTextStyle: { textColor: theme.color.textNeutral },
    mode: {
      disabled: {
        detailsTextStyle: { textColor: theme.color.textNeutralDisabled }
      }
    }
  }
});
