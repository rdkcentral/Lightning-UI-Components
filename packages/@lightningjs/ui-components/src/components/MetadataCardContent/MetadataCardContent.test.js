/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';
import MetadataCardContent from '.';
import xfinityProviderLogoSquare from '../../assets/images/Xfinity-Provider-Logo-Square.png';

const createComponent = makeCreateComponent(MetadataCardContent);

describe('MetadataCardContent', () => {
  let metadataCardContent, testRenderer;

  beforeEach(() => {
    [metadataCardContent, testRenderer] = createComponent(
      {
        provider: {
          visibleCount: 3,
          providers: Array(10).fill(xfinityProviderLogoSquare)
        }
      },
      {
        spyOnMethods: ['_update']
      }
    );
  });
  afterEach(() => {
    metadataCardContent = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the appropriate text content status', () => {
    const title = 'Title';
    const description = 'Description';
    const descriptionDetails = 'Description Details';
    const details = 'Details';
    const provider = { providers: [{ icon: 'test.png', announce: 'test' }] };
    metadataCardContent.patch({
      title,
      description,
      descriptionDetails,
      details,
      provider
    });
    testRenderer.forceAllUpdates();
    expect(metadataCardContent.announce).toEqual([
      title,
      description,
      descriptionDetails,
      details,
      ['test']
    ]);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    metadataCardContent.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(metadataCardContent.announce).toBe(overrideString);
  });

  it('updates the title', async () => {
    const title = 'title text';
    expect(metadataCardContent.title).toBe(undefined);
    metadataCardContent.title = title;
    await metadataCardContent.__updateSpyPromise;
    expect(metadataCardContent._Title.content).toBe(title);
  });

  it('updates the subtitle', async () => {
    const subtitle = 'subtile text';
    expect(metadataCardContent.subtitle).toBe(undefined);
    metadataCardContent.subtitle = subtitle;
    await metadataCardContent.__updateSpyPromise;
    expect(metadataCardContent._Subtitle.content).toBe(subtitle);
  });

  it('updates the description', async () => {
    const description = 'description text';
    expect(metadataCardContent.description).toBe(undefined);
    metadataCardContent.description = description;
    await metadataCardContent.__updateSpyPromise;
    expect(metadataCardContent._Description.content).toBe(description);
  });

  it('updates the description details', async () => {
    const descriptionDetails = 'descriptionDetails text';
    expect(metadataCardContent.descriptionDetails).toBe(undefined);
    metadataCardContent.descriptionDetails = descriptionDetails;
    await metadataCardContent.__updateSpyPromise;
    expect(metadataCardContent._DescriptionDetails.content).toBe(
      descriptionDetails
    );
  });

  it('updates the details', async () => {
    const details = 'details text';
    expect(metadataCardContent.details).toBe(undefined);
    metadataCardContent.details = details;
    await metadataCardContent.__updateSpyPromise;
    expect(metadataCardContent._Details.content).toBe(details);
  });

  it('creates provider', () => {
    expect(
      metadataCardContent._Provider._Row.items[3].tag('Text').content
    ).toBe('+7');
    expect(metadataCardContent._Provider._Row._Items.h).toBe(
      metadataCardContent.style.provider.itemSize
    );
  });
});
