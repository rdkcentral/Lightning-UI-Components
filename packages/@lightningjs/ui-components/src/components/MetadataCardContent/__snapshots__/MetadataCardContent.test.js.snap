// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataCardContent renders 1`] = `
{
  "Component": {
    "DetailsWrapper": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "DetailsFader": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Details": {
              "active": false,
              "alpha": 0.001,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0.5,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Details",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "Details",
              ],
              "type": "TextBox",
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 30,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "DetailsFader",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "DetailsFader",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Provider": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Row": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Items": {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "children": {
                    "Element-13": {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294506490,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 60,
                      "hasFinalFocus": false,
                      "hasFocus": false,
                      "isComponent": true,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": "",
                      "tag": [Function],
                      "type": "Icon",
                      "visible": true,
                      "w": 60,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-14": {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294506490,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 60,
                      "hasFinalFocus": false,
                      "hasFocus": false,
                      "isComponent": true,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": "",
                      "tag": [Function],
                      "type": "Icon",
                      "visible": true,
                      "w": 60,
                      "x": 70,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-15": {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294506490,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 60,
                      "hasFinalFocus": false,
                      "hasFocus": false,
                      "isComponent": true,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": "",
                      "tag": [Function],
                      "type": "Icon",
                      "visible": true,
                      "w": 60,
                      "x": 140,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-16": {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "children": {
                        "Background": {
                          "active": false,
                          "alpha": 1,
                          "attached": true,
                          "boundsMargin": null,
                          "clipping": false,
                          "color": 4294967295,
                          "enabled": true,
                          "flex": false,
                          "flexItem": false,
                          "h": 60,
                          "isComponent": undefined,
                          "mount": 0,
                          "mountX": 0,
                          "mountY": 0,
                          "pivot": 0.5,
                          "pivotX": 0.5,
                          "pivotY": 0.5,
                          "ref": "Background",
                          "renderOfScreen": undefined,
                          "renderToTexture": false,
                          "scale": 1,
                          "scaleX": 1,
                          "scaleY": 1,
                          "state": undefined,
                          "tag": [Function],
                          "tags": [
                            "Background",
                          ],
                          "texture": {
                            "type": "StaticCanvasTexture",
                          },
                          "visible": true,
                          "w": 60,
                          "x": 0,
                          "y": 0,
                          "zIndex": 0,
                        },
                        "Text": {
                          "active": false,
                          "alpha": 0.001,
                          "attached": true,
                          "boundsMargin": null,
                          "children": {
                            "Text": {
                              "active": false,
                              "alpha": 1,
                              "attached": true,
                              "boundsMargin": null,
                              "clipping": false,
                              "color": 4294967295,
                              "enabled": true,
                              "flex": false,
                              "flexItem": false,
                              "h": 0,
                              "isComponent": undefined,
                              "mount": 0,
                              "mountX": 0,
                              "mountY": 0,
                              "pivot": 0.5,
                              "pivotX": 0.5,
                              "pivotY": 0.5,
                              "ref": "Text",
                              "renderOfScreen": undefined,
                              "renderToTexture": false,
                              "scale": 1,
                              "scaleX": 1,
                              "scaleY": 1,
                              "state": undefined,
                              "tag": [Function],
                              "tags": [
                                "Text",
                              ],
                              "texture": {
                                "fontFace": "Arial",
                                "fontSize": 25,
                                "fontStyle": "500",
                                "lineHeight": 36,
                                "precision": 0.6666666666666666,
                                "text": "+7",
                                "textBaseline": "bottom",
                                "textColor": 4294506490,
                                "type": "TextTexture",
                                "verticalAlign": "middle",
                              },
                              "visible": true,
                              "w": 0,
                              "x": 0,
                              "y": 2,
                              "zIndex": 0,
                            },
                          },
                          "clipping": false,
                          "color": 4294967295,
                          "enabled": true,
                          "flex": false,
                          "flexItem": false,
                          "h": 0,
                          "hasFinalFocus": false,
                          "hasFocus": false,
                          "isComponent": true,
                          "mount": 0.5,
                          "mountX": 0.5,
                          "mountY": 0.5,
                          "pivot": 0.5,
                          "pivotX": 0.5,
                          "pivotY": 0.5,
                          "ref": "Text",
                          "renderOfScreen": undefined,
                          "renderToTexture": false,
                          "scale": 1,
                          "scaleX": 1,
                          "scaleY": 1,
                          "state": "",
                          "tag": [Function],
                          "tags": [
                            "Text",
                          ],
                          "type": "TextBox",
                          "visible": true,
                          "w": 0,
                          "x": 30,
                          "y": 30,
                          "zIndex": 0,
                        },
                      },
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 60,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "visible": true,
                      "w": 60,
                      "x": 210,
                      "y": 0,
                      "zIndex": 0,
                    },
                  },
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 60,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Items",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Items",
                  ],
                  "visible": true,
                  "w": 270,
                  "x": 0,
                  "y": 0,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 60,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Row",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "Row",
              "tag": [Function],
              "tags": [
                "Row",
              ],
              "type": "Row",
              "visible": true,
              "w": 270,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 1,
          "mountX": 1,
          "mountY": 1,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Provider",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Provider",
          ],
          "type": "Provider",
          "visible": true,
          "w": 270,
          "x": -270,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 60,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "DetailsWrapper",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "DetailsWrapper",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": -60,
      "zIndex": 0,
    },
    "Text": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Description": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Description",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Description",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "DescriptionDetails": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "DescriptionDetails",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "DescriptionDetails",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 5,
          "zIndex": 0,
        },
        "Title": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Title",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Title",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Text",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Text",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
